GitExtensions/StaticDependencyInjection.cs

```diff
@@ -0,0 +1,118 @@
+﻿using GitCommands;
+using System;
+using System.IO;
+using System.Collections.Concurrent;
+
+namespace GitExtensions
+{
+
+    public class DIObjectInstance<T> where T : class, new()
+    {
+        private static object lockObj = new object();
+        private T _Inst;
+        public T Inst
+        {
+            get
+            {
+                if (_Inst == null)
+                {
+                    lock (lockObj)
+                    {
+                        if (_Inst == null)
+                        {
+                            _Inst = StaticDI.CreateInstance<T>(ClearInst);
+                        }
+                    }
+                }
+
+                return _Inst;
+            }
+        }
+
+        private void ClearInst()
+        {
+            lock (lockObj)
+            {
+                _Inst = null;
+            }
+        }
+    }
+
+    public interface IInstanceFactory
+    {
+        T CreateInstance<T>() where T : class, new();
+    }
+
+    public class NewInstanceFactory : IInstanceFactory
+    {
+        public T CreateInstance<T>() where T : class, new()
+        {
+            return new T();
+        }
+    }
+
+    public class StaticDI
```

Comment:
>It feels we're reinventing the wheel here. Why not use Autofac or alike?

Sure, we are:). I don't want to reinvent the wheel. I just wanted to express the idea but didn't want to learn a new library to do that. I wanted to discuss the general idea of using global instances vs passing them through a constructor.